File: 988/988.ts
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/988.py

function frogSum(aIn: number, bIn: number): bigint {
    if (aIn <= 0 || bIn <= 0) {
        throw new Error("a and b must be positive");
    }
    let a = aIn, b = bIn;
    if (a > b) {
        [a, b] = [b, a];
    }
    if (a === 1) {
        return 0n;
    }
    const aBig = BigInt(a);
    const bBig = BigInt(b);
    const width = bBig - 1n;
    const h: bigint[] = new Array(Number(width) + 1).fill(0n);
    for (let i = 1; i <= Number(width); i++) {
        h[i] = (aBig * bBig - aBig * BigInt(i) - 1n) / bBig;
    }
    let dp: Map<bigint, [bigint, bigint]> = new Map();
    for (let t = 0n; t <= h[1]; t++) {
        dp.set(t, [1n, 0n]);
    }
    for (let i = 2; i <= Number(width); i++) {
        const nextDp: Map<bigint, [bigint, bigint]> = new Map();
        for (const [prevHeight, [count, total]] of dp.entries()) {
            const limit = prevHeight < h[i] ? prevHeight : h[i];
            for (let curHeight = 0n; curHeight <= limit; curHeight++) {
                let add = 0n;
                if (prevHeight > curHeight && prevHeight > 0n) {
                    add = aBig * bBig - aBig * (BigInt(i) - 1n) - bBig * prevHeight;
                }
                const existing = nextDp.get(curHeight) || [0n, 0n];
                nextDp.set(curHeight, [
                    existing[0] + count,
                    existing[1] + total + count * add,
                ]);
            }
        }
        dp = nextDp;
    }
    let answer = 0n;
    const lastColumn = width;
    for (const [prevHeight, [count, total]] of dp.entries()) {
        let add = 0n;
        if (prevHeight > 0n) {
            add = aBig * bBig - aBig * lastColumn - bBig * prevHeight;
        }
        answer += total + count * add;
    }
    return answer;
}

function solve(): void {
    console.assert(frogSum(3, 5) === 23n, "frogSum(3,5) should be 23");
    console.assert(frogSum(5, 13) === 16336n, "frogSum(5,13) should be 16336");
    console.log(frogSum(19, 53).toString());
}

solve();

File: 988/988.cs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/988.py

using System;
using System.Collections.Generic;

class Program
{
    static long FrogSum(long aIn, long bIn)
    {
        if (aIn <= 0 || bIn <= 0)
            throw new ArgumentException("a and b must be positive");
        long a = aIn, b = bIn;
        if (a > b) { long tmp = a; a = b; b = tmp; }
        if (a == 1) return 0;
        long width = b - 1;
        long[] h = new long[width + 1];
        for (long i = 1; i <= width; i++)
            h[i] = (a * b - a * i - 1) / b;
        var dp = new Dictionary<long, (long count, long total)>();
        for (long t = 0; t <= h[1]; t++)
            dp[t] = (1, 0);
        for (long i = 2; i <= width; i++)
        {
            var nextDp = new Dictionary<long, (long count, long total)>();
            foreach (var kvp in dp)
            {
                long prevHeight = kvp.Key;
                long count = kvp.Value.count;
                long total = kvp.Value.total;
                long limit = Math.Min(prevHeight, h[i]);
                for (long curHeight = 0; curHeight <= limit; curHeight++)
                {
                    long add = 0;
                    if (prevHeight > curHeight && prevHeight > 0)
                        add = a * b - a * (i - 1) - b * prevHeight;
                    if (!nextDp.TryGetValue(curHeight, out var existing))
                        existing = (0, 0);
                    nextDp[curHeight] = (
                        existing.count + count,
                        existing.total + total + count * add
                    );
                }
            }
            dp = nextDp;
        }
        long answer = 0;
        long lastColumn = width;
        foreach (var kvp in dp)
        {
            long prevHeight = kvp.Key;
            long count = kvp.Value.count;
            long total = kvp.Value.total;
            long add = 0;
            if (prevHeight > 0)
                add = a * b - a * lastColumn - b * prevHeight;
            answer += total + count * add;
        }
        return answer;
    }

    static void Solve()
    {
        System.Diagnostics.Debug.Assert(FrogSum(3, 5) == 23);
        System.Diagnostics.Debug.Assert(FrogSum(5, 13) == 16336);
        Console.WriteLine(FrogSum(19, 53));
    }

    static void Main()
    {
        Solve();
    }
}
